src/Orchard.Web/Modules/Orchard.OpenId/Services/AzureActiveDirectory/InMemoryCache.cs

```diff
@@ -1,39 +1,51 @@
 ﻿using System;
 using System.Collections.Concurrent;
-using System.Collections.Generic;
-using System.Linq;
 using Microsoft.IdentityModel.Clients.ActiveDirectory;
+using Orchard.Environment.Extensions;
 
-namespace Orchard.OpenId {
-    public class InMemoryCache : TokenCache {
-        private const string CacheIdSuffix = "_TokenCache";
-        private static ConcurrentDictionary<string, byte[]> _inMemoryTokenCache;
-
-        private string _userObjectId;
-        private string _cacheId;
-
-        public InMemoryCache(string userId) {
-            _userObjectId = userId;
-            _cacheId = String.Concat(_userObjectId, CacheIdSuffix);
-
-            if (_inMemoryTokenCache == null)
-                _inMemoryTokenCache = new ConcurrentDictionary<string, byte[]>();
+namespace Orchard.OpenId.Services.AzureActiveDirectory {
+    [OrchardFeature("Orchard.OpenId.AzureActiveDirectory")]
+    public class InMemoryCache : TokenCache, ISingletonDependency {
+        public InMemoryCache() {
+            // On tenant start clear cache.
+            Clear();
```

Comment:
Not really necessary as the instance is new on tenant restart. Or there is an issue with the design.